relay-server/src/services/project_local.rs

```diff
@@ -95,30 +98,30 @@ async fn load_local_states(
         }
 
         // serde_json is not async, so spawn a blocking task here:
-        let (path, state) = tokio::task::spawn_blocking(move || parse_file(path)).await??;
+        let (path, mut state) = tokio::task::spawn_blocking(move || parse_file(path)).await??;
 
-        let mut sanitized = ProjectState::sanitize(state);
-        if sanitized.project_id.is_none() {
+        if state.info.project_id.is_none() {
             if let Some(project_id) = get_project_id(&path) {
-                sanitized.project_id = Some(project_id);
+                state.info.project_id = Some(project_id);
             } else {
                 relay_log::warn!(?path, "skipping file, filename is not a valid project id");
                 continue;
             }
         }
 
         // Keep a separate project state per key.
-        let keys = std::mem::take(&mut sanitized.public_keys);
+        let keys = std::mem::take(&mut state.info.public_keys);
         for key in keys {
-            sanitized.public_keys = smallvec::smallvec![key.clone()];
-            states.insert(key.public_key, Arc::new(sanitized.clone()));
+            let mut state = state.clone();
+            state.info.public_keys = smallvec::smallvec![key.clone()];
+            states.insert(key.public_key, ProjectState::from(state).sanitize());
```

Comment:
This makes it hard to make `ParsedProjectState` a private concern of the (de)serializer: We need to peek into the config to get the project key.